### DISCUSSION TITLE
Rotating Anon, Service, and JWT Secrets

### DISCUSSION THREAD

**Discussion by tylerfontaine** [supabase/supabase#20031] [Troubleshooting]
_This is a copy of a troubleshooting article on Supabase's docs site. It may be missing some details from the original. View the [original article](https://supabase.com/docs/guides/troubleshooting/rotating-anon-service-and-jwt-secrets-1Jq6yd)._

Have you ever accidentally committed a service key to a public repo? Or maybe rotating keys is something you regularly do for security compliance.
Whatever the reason, here's how to rotate the keys for your Supabase project.

If you haven’t migrated to asymmetric JWT signing keys:

We recommend that you migrate to asymmetric JWT signing keys and publishable/secret API keys as it is no longer possible to rotate the legacy anon, service and JWT secrets.
You can view this [**Get Started guide**](https://supabase.com/docs/guides/auth/signing-keys#getting-started) for steps to migrate to asymmetric JWT signing keys.

If you have migrated to new symmetric JWT signing keys:

1.  Go to [**Project Settings** → **JWT Keys**](https://supabase.com/dashboard/project/_/settings/jwt)s in the Supabase Dashboard
2.  Navigate to the **JWT Signing Keys** tab.
3.  Click on Rotate Keys. This will move the current key to “Previously used keys”
4.  Select the three-dot icon (action icon) of your previously used key and click “Revoke”. If you do not “Revoke” the key, older keys will still be valid.

## Further readings

*   [Rotate a leaked or compromised key](https://supabase.com/docs/guides/getting-started/api-keys#leaked-key)
*   [JWT Signing Keys](https://supabase.com/docs/guides/auth/signing-keys)

**Comment by rayyeoh** [comment]
I've generated a new JWT secret but it has been 4 hours since. How long is the next step suppose to take? Do I restart  the project whilst still in this state or has something gone wrong? Thank you for your input.

![jwtrefresh](https://github.com/supabase/supabase/assets/57467750/736a5644-55fe-4f56-895d-89a0f5057717)

**Comment by GregTillman-coder** [comment]
i've done exactly this but the regenerate button is gone and they are pushing new JWT keys on me?! i need to fix this quick and can't change everything else i just need new keys
